### PR TITLE
feat(): use adapters

### PR DESCRIPTION
**Pull request by TylorS**
Uses an adapter interface to allow for cycle-core to adapt
to any and all stream/observable libraries that have an adapter
written. This branch uses an Rx 4 adapter for the tests.

### CODE REVIEW

**Comment by Frikki** [comment]
:+1: Very clever. Code of my own heart. I look forward testing this out. :+1: again!

**Comment by TylorS** [comment]
With this cycle-core can be entirely stream/observable library agnostic via adapters. 
Currently I have only written an Rx 4 adapter https://github.com/tylors/cycle-rx-adapter
The way this works currently.
```js
import Cycle from '@cycle/core'
import streamAdapter from 'cycle-rx-adapter'

...code here....

run(main, driver, {streamAdapter})
```

A driver that is **not** written in the library that the streamAdapater you chose to give to `run()`
can be coerced into working by adding a streamAdapter property to the driver function.

```js
function driverFn(sink$) {...}

driverFn.streamAdapter = rxAdapter // mostAdapter 
```

**Comment by staltz** [comment]
:+1: for the adapter solution in general.

About the API:
Because "just using Cycle with Rx" is going to be far more common than "I want to use Cycle with my own non-officially-supported streams library", an API like
```js
import Rx from 'rx'

run(main, driver, {streamLibrary: Rx})
```
would be better. In fact, we can assume Rx as the default library. So then we can change package.json

```diff
-"cycle-rx-adapter": "^0.2.0",
 "rx": "4.0.7",
```

It'd make more sense for most Cycle developers to install `npm install rx @cycle/core @cycle/dom` than to install `npm install rx cycle-rx-adapter @cycle/core @cycle/dom`.

Also, in the streams adapter code, I suppose `sink` is now an overloaded term. We should probably use `observer`.

**Comment by TylorS** [comment]
I'm getting ready for sleep, so I'll have to try to get a better reply later, but personally I have no plans to ever use Rx. I plan to use *Most* only, and I don't want to have Rx in my code base if it's not required. That would be required with what you propose. My goals with this is to allow cycle-core to become a single place for both visions (and more) to exist. It's fair to say I'm probably on the outside looking in there.

It's also possible to wrap this up in other libraries with default adapters if that is desired.

```shell
npm install @cycle/rx @cycle/dom
npm install @cycle/most @cycle/dom
npm install @cycle/whatever @cycle/dom
```

At the end of the day, this isn't my library though.

**Comment by staltz** [comment]
> I'm getting ready for sleep

Oh yeah, forgot the timezone issues!

**Comment by TylorS** [comment]
Been up about 27 hours haha

**Comment by staltz** [comment]
> personally I have no plans to ever use Rx. I plan to use Most only, and I don't want to have Rx in my code base if it's not required

Ok, that affects only this:

> In fact, we can assume Rx as the default library

How bad would it be to publish packages
- `@cycle/core-rx`
- `@cycle/core-most`
- `@cycle/core-rx5`
- `@cycle/core-bacon`

To replace the `{streamLibrary}`-type API? Then we can freely determine that `@cycle/core` is an alias to `@cycle/core-rx`, and later on an alias for `@cycle/core-rx5`. It would mean writing Cycle Core over and over a few times, but it's small, and the end goal is to have a good API for users anyway. And in your case you would only have to use `@cycle/core-most`, no attachment to Rx.

**Comment by staltz** [comment]
Maybe that's what you meant with `npm install @cycle/rx`?

**Comment by staltz** [comment]
And again I'm caught thinking of bundling the stream library inside `@cycle/core-the-stream-library` if we are going to do the above API.

**Comment by TylorS** [comment]
`@cycle/rx`
```js
import Cycle from '@cycle/core'
import streamAdapter from 'cycle-rx-adapter'

const cycle = {
  run: (main, drivers) => Cycle.run(main, drivers, {streamAdapter})
}

export default cycle
```

**Comment by Cmdv** [comment]
this does seem the most intuitive to me `npm install @cycle/rx` :+1: 
Got to add really liking how this idea has bought everything back to a single point. Not a single atom of course :P

**Comment by staltz** [comment]
In that case we could rename `core` to `base`, since it's not meant to be directly used by users. As for the naming, I can see people thinking that `@cycle/rx` is a bundle of Rx library. We could name it instead `@cycle/rx-run`. "*Run() for RxJS*"

**Comment by TylorS** [comment]
Should 'officially' supported adapters exist within `@cycle/` as well? 
Likely just rx4 rx5 and *most*

**Comment by TylorS** [comment]
Other than that, I find naming to be trivial.

**Comment by staltz** [comment]
Yeah I think `@cycle/most-adapter` makes sense

**Comment by TylorS** [comment]
If we rename `@cycle/core` to `@cycle/base`
Should we then create `@cycle/core` as an 'alias' to Rx as default? In theory 0 changes would be required by the end user.

**Comment by staltz** [comment]
Good point. We can do the alias `@cycle/core` => `@cycle/rx-run` but socially deprecate `@cycle/core`, encouraging people to use `@cycle/rx-run` (just the package name is different). Meaning that `core` is an older major version thingy.

**Comment by TylorS** [comment]
That sounds appropriate. :+1:

**Comment by staltz** [comment]
We can also alias `@cycle/TylorS` => `@cycle/sleep`

**Comment by TylorS** [comment]
I'm inclined to suggest `@cycle/TylorS` to be the greatest package name of all time. No bias at all.

## Loose list of things to do

- `@cycle/core` -> `@cycle/base`
- `@cycle/core` = `@cycle/rx-run` = `@cycle/base + @cycle/rx-adapter`
- deprecate `@cycle/core`
- Update all cycle drivers to use `driverFn.streamAdapter = rxAdapter`
- Write Most.js adapter
- create `@cycle/most-run`
- Update motorcycle drivers to use `driverFn.streamAdapter = mostAdapter`

Anything else?

Should we move motorcycle-dom to `@cycle/snabbdom` or similar once it's up to date with the recent `@cycle/dom` changes regarding event delegation?

**Comment by staltz** [comment]
Everything correct.

And one thing just to be sure: *add test on `@cycle/base` to make sure one-liner drivers use the same stream library as the run() is using, by default*.

**Comment by TylorS** [comment]
Updated

**Comment by TylorS** [comment]
https://github.com/TylorS/cycle-most-adapter

File: src/cycle.js [hunk 1]
```diff
@@ -1,88 +1,55 @@
...
+        drivers[name].streamAdapter || runStreamAdapter
 
-function attachDisposeToSinks(sinks, replicationSubscription) {
-  Object.defineProperty(sinks, `dispose`, {
-    enumerable: false,
-    value: () => { replicationSubscription.dispose() },
-  })
-  return sinks
-}
+      const adapt = stream => // eslint-disable-line
```
Comment (Frikki) [inline src/cycle.js]:
```js
const streamAdaptation = ...
```

File: src/cycle.js [hunk 2]
```diff
@@ -1,88 +1,55 @@
...
 
-function makeDisposeSources(sources) {
-  return function dispose() {
-    for (let name in sources) {
-      if (sources.hasOwnProperty(name) &&
-        typeof sources[name].dispose === `function`)
-      {
-        sources[name].dispose()
-      }
+      sources[name] = drivers[name](sinkProxies[name].stream, adapt, name)
```
Comment (Frikki) [inline src/cycle.js]:
```js
sources[name] = drivers[name](sinkProxies[name].stream, streamAdaptationFunc, name)
```

**Comment by staltz** [comment]
@TylorS I got your list of things to do, ordered them in a particular order, and put it in #196, since that's a better place for a to do list.

**Comment by Frikki** [comment]
> Because "just using Cycle with Rx" is going to be far more common than "I want to use Cycle with my own non-officially-supported streams library", an API like ... would be better. In fact, we can assume Rx as the default library.

By same logic, *@cycle/dom* should also then be *default*, because using it is going to be far more common than "I want to use Cycle with my own non-officially-supported DOM library".

The problem I see with *Rx* as default is that *Rx* will be included in my app code whether I use it or not.

In my view, if you want to provide users with a default setup, just make a new repo/package that has the necessary dependencies and imports to your liking. This will leave other repo/packages separated with their respective concerns. (I hear the crowd cheering for separation of concerns and the flexibility it provides).

**Comment by staltz** [comment]
@Frikki that comment was "deprecated". Remove the last sentence:

```diff
 Because "just using Cycle with Rx" is going to be far more common 
 than "I want to use Cycle with my own non-officially-supported streams 
 library", an API like ... would be better. 
-In fact, we can assume Rx as the default library.
```

**Comment by Frikki** [comment]
@staltz Okay. Is the Rx dependency solved than? I find it a bit difficult to follow the above discussion.

**Comment by staltz** [comment]
It is solved... I updated issue #196 summarizing our solution.

**Comment by TylorS** [comment]
Closing this as we're moving this PR to [base](https://github.com/cyclejs/base)

**Pull request closed without merging**